torchao/csrc/cuda/fp6_llm/fp6_linear.cu

```diff
@@ -18,8 +18,6 @@
 // - Modified the TilingConfig parameters for SM75 to deal with smaller shared memory
 //
 
-#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 750  // at least Turing
```

Comment:
I think we still need to keep this. Although it seems like this file only contains host code, `fp6_linear.cu` will include `kernel_matmul.cuh` and `kernel_reduction.cuh` (which in turns include more header files), which has device code. I rmb adding this line in #636.

So what I think is happening on old devices (CUDA_ARCH < 7.5)
1. Host code is still compiled (in fact it is always compiled) e.g. `fpx_linear_kernel()`
2. Device code will not be compiled e.g. `CopyFromGlobalToShared_A()`
3. When trying to run the host function (`fpx_linear_kernel()`), you will get undefined symbol errors, which are the device code functions that were not compiled.

Without this guard, when you try to compile this file on CUDA_ARCH < 7.5, you will get undefined nv_bfloat16 symbol (or something along that line) I think. Didn't check my hypothesis, but that's what I'm suspecting is happening.

Reply:
I just tested on SM75 and set the guard at the top of `fp6_linear.cu` to `#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 800` specifically to test what would happen for an unsupported version (after verifying that the kernel does work correctly for SM75, see results below). The expected outcome would be that the code throws an error complaining about undefined symbols, but surprisingly, the code still compiles and runs. Earlier, this led me to believe that the guard does not do anything, but this doesn't seem to be what's happening. You're right that the guard does have an effect, but instead of failing, the kernel then returns incorrect results and has a constant runtime no matter the input shape (see results below). This seems to indicate that the kernel returns silently for unsupported SM versions instead of the expected outcome of erroring out.

Would you agree this is problematic? E.g. if someone compiles for SM70 and runs FP6 than the code will still work, but it will return garbage outputs. From this perspective, the guard actually seems to do more harm than good, since it gives the impression that the code runs correctly, even though it might not. So then leaving out the guard would at least catch some arch-specific errors (e.g. `__nv_bfloat16` not being defined).

A possible other solution would be to leave the guard in and check for sm>=75 at runtime (e.g. in the Python code that calls the FP6 kernel). Then the codebase can still compile for older architectures, while still making sure that the FP6 kernel cannot be executed. What do you think?

**Case 1**
```cpp
#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 750  // tested on sm75
```
![image](https://github.com/user-attachments/assets/44e7f216-3dcb-4ab7-a42a-6fe61af489ef)


**Case 2**
```cpp
#if !defined(__CUDA_ARCH__) || __CUDA_ARCH__ >= 800  // tested on sm75
```
![image](https://github.com/user-attachments/assets/70a3f3fe-b3b0-4cf9-9b70-aa8fda5666d9)

Reply:
That's.. interesting. I rmb we had an error before when trying to compile on old hardware #632, hence I added the CUDA_ARCH guard, and it compiles. Now thinking about it, I didn't test what would happen when I call the function on said old hardware, which would output garbage results, but no error, like you discovered.

I was following this https://github.com/pytorch/pytorch/blob/e98eac76b358fb4639b9e9ce6894014354d7b073/aten/src/ATen/native/cuda/int4mm.cu#L1 for the compile guard. And running tinygemm (this int4 mm kernel) on unsupported hardware will result in undefined symbol error at runtime. You can run this to reproduce on google colab https://github.com/pytorch/ao/issues/1110#issuecomment-2428090812.

I don't exactly like sm>=75 check at runtime (well, we already do that for sm75 launch params). I will try to play around with these compile guards to see their behavior. Now it seems like trying to support sm75 is getting very messy 🤣.

Reply:
@tobiasvanderwerff I created a thread over at GPU-MODE discord for easier discussion (writing it here in case you don't frequent discord much)